File: backend/app/schemas/templates.py
from pydantic import BaseModel, Field
from typing import Dict, List, Optional, Any, Union

from .shared_schemas import ProjectTechStack, Requirements


class FeatureModule(BaseModel):
    """Feature module configuration."""

    name: str
    description: str
    enabled: bool = True
    optional: bool = False
    providers: List[str] = Field(default_factory=list)


class Features(BaseModel):
    """Features configuration."""

    coreModules: List[FeatureModule] = Field(default_factory=list, alias="coreModules")

    class Config:
        populate_by_name = True


class PageComponent(BaseModel):
    """Page component."""

    name: str
    path: str
    components: List[str] = Field(default_factory=list)
    enabled: bool = True


class Pages(BaseModel):
    """Application pages."""

    public: List[PageComponent] = Field(default_factory=list)
    authenticated: List[PageComponent] = Field(default_factory=list)
    admin: List[PageComponent] = Field(default_factory=list)


class EntityField(BaseModel):
    """Data model entity field."""

    name: str
    type: str
    primary_key: bool = False
    generated: bool = False
    unique: bool = False
    required: bool = False
    default: Optional[Union[str, int, float, bool]] = None
    enum: Optional[List[str]] = None
    foreign_key: Optional[Dict[str, Any]] = None


class Entity(BaseModel):
    """Data model entity."""

    name: str
    description: str
    fields: List[EntityField] = Field(default_factory=list)


class Relationship(BaseModel):
    """Data model relationship."""

    type: str
    from_entity: str
    to_entity: str
    field: str

    class Config:
        populate_by_name = True


class DataModel(BaseModel):
    """Application data model."""

    entities: List[Entity] = Field(default_factory=list)
    relationships: List[Relationship] = Field(default_factory=list)

    class Config:
        populate_by_name = True


class ApiEndpoint(BaseModel):
    """API endpoint configuration."""

    path: str
    description: str
    methods: List[str] = Field(default_factory=list)
    auth: bool = False
    roles: List[str] = Field(default_factory=list)


class Api(BaseModel):
    """API configuration."""

    endpoints: List[ApiEndpoint] = Field(default_factory=list)


class TestingFramework(BaseModel):
    """Testing framework configuration."""

    framework: str
    coverage: int = 0
    directories: Optional[List[str]] = None
    focus: Optional[List[str]] = None
    scenarios: Optional[List[str]] = None


class Testing(BaseModel):
    """Testing configuration."""

    unit: TestingFramework
    integration: TestingFramework
    e2e: TestingFramework


class ProjectStructure(BaseModel):
    """Project structure configuration."""

    frontend: Dict[str, List[str]]


class DeploymentEnvironment(BaseModel):
    """Deployment environment configuration."""

    name: str
    url: str
    variables: List[Dict[str, Any]] = Field(default_factory=list)


class CICD(BaseModel):
    """CI/CD configuration."""

    provider: str
    steps: List[str] = Field(default_factory=list)


class Deployment(BaseModel):
    """Deployment configuration."""

    environments: List[DeploymentEnvironment] = Field(default_factory=list)
    cicd: CICD


class Diagram(BaseModel):
    """Documentation diagram."""

    name: str
    type: str
    template: str


class Documentation(BaseModel):
    """Documentation configuration."""

    specs: List[str] = Field(default_factory=list)
    diagrams: List[Diagram] = Field(default_factory=list)


class Colors(BaseModel):
    """Colors schema for UI design."""

    primary: str = "#3b82f6"
    secondary: str = "#6366f1"
    accent: str = "#f59e0b"
    background: str = "#ffffff"
    text_primary: str = Field(alias="textPrimary", default="#1f2937")
    text_secondary: str = Field(alias="textSecondary", default="#4b5563")
    success: str = "#10b981"
    warning: str = "#f59e0b"
    error: str = "#ef4444"
    info: str = "#3b82f6"
    surface: str = "#ffffff"
    border: str = "#e5e7eb"

    class Config:
        populate_by_name = True


class HeadingSizes(BaseModel):
    """Heading sizes schema for typography."""

    h1: str = "2.5rem"
    h2: str = "2rem"
    h3: str = "1.75rem"
    h4: str = "1.5rem"
    h5: str = "1.25rem"
    h6: str = "1rem"

    class Config:
        populate_by_name = True


class Typography(BaseModel):
    """Typography schema for UI design."""

    font_family: str = Field(alias="fontFamily", default="Inter, sans-serif")
    heading_font: str = Field(alias="headingFont", default="Inter, sans-serif")
    font_size: str = Field(alias="fontSize", default="16px")
    line_height: float = Field(alias="lineHeight", default=1.5)
    font_weight: int = Field(alias="fontWeight", default=400)
    heading_sizes: HeadingSizes = Field(alias="headingSizes", default_factory=HeadingSizes)

    class Config:
        populate_by_name = True


class Spacing(BaseModel):
    """Spacing schema for UI design."""

    unit: str = "4px"
    scale: List[int] = [0, 1, 2, 3, 4, 5, 6, 8, 10, 12, 16, 20, 24, 32, 40, 48, 64, 80, 96]

    class Config:
        populate_by_name = True


class BorderRadius(BaseModel):
    """Border radius schema for UI design."""

    small: str = "2px"
    medium: str = "4px"
    large: str = "8px"
    xl: str = "12px"
    pill: str = "9999px"

    class Config:
        populate_by_name = True


class Shadows(BaseModel):
    """Shadows schema for UI design."""

    small: str = "0 1px 2px 0 rgba(0, 0, 0, 0.05)"
    medium: str = "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)"
    large: str = "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)"
    xl: str = "0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)"

    class Config:
        populate_by_name = True


class Layout(BaseModel):
    """Layout schema for UI design."""

    container_width: str = Field(alias="containerWidth", default="1280px")
    responsive: bool = True
    sidebar_width: str = Field(alias="sidebarWidth", default="240px")
    topbar_height: str = Field(alias="topbarHeight", default="64px")
    grid_columns: int = Field(alias="gridColumns", default=12)
    gutter_width: str = Field(alias="gutterWidth", default="16px")

    class Config:
        populate_by_name = True


class Components(BaseModel):
    """Components schema for UI design."""

    button_style: str = Field(alias="buttonStyle", default="rounded")
    input_style: str = Field(alias="inputStyle", default="outline")
    card_style: str = Field(alias="cardStyle", default="shadow")
    table_style: str = Field(alias="tableStyle", default="bordered")
    nav_style: str = Field(alias="navStyle", default="pills")

    class Config:
        populate_by_name = True


class DarkModeColors(BaseModel):
    """Dark mode colors schema."""

    background: str = "#1f2937"
    text_primary: str = Field(alias="textPrimary", default="#f9fafb")
    text_secondary: str = Field(alias="textSecondary", default="#e5e7eb")
    surface: str = "#374151"
    border: str = "#4b5563"

    class Config:
        populate_by_name = True


class DarkMode(BaseModel):
    """Dark mode schema."""

    enabled: bool = True
    colors: DarkModeColors = Field(default_factory=DarkModeColors)

    class Config:
        populate_by_name = True


class Animations(BaseModel):
    """Animations schema."""

    transition_duration: str = Field(alias="transitionDuration", default="150ms")
    transition_timing: str = Field(alias="transitionTiming", default="ease-in-out")
    hover_scale: float = Field(alias="hoverScale", default=1.05)
    enable_animations: bool = Field(alias="enableAnimations", default=True)

    class Config:
        populate_by_name = True


class UIDesign(BaseModel):
    """UI design schema."""

    colors: Colors = Field(default_factory=Colors)
    typography: Typography = Field(default_factory=Typography)
    spacing: Spacing = Field(default_factory=Spacing)
    border_radius: BorderRadius = Field(alias="borderRadius", default_factory=BorderRadius)
    shadows: Shadows = Field(default_factory=Shadows)
    layout: Layout = Field(default_factory=Layout)
    components: Components = Field(default_factory=Components)
    dark_mode: DarkMode = Field(alias="darkMode", default_factory=DarkMode)
    animations: Animations = Field(default_factory=Animations)

    class Config:
        populate_by_name = True


class ProjectTemplate(BaseModel):
    """Project template schema."""

    name: str
    version: str
    description: str
    business_goals: List[str] = Field(default_factory=list, alias="businessGoals")
    target_users: str = Field(default="", alias="targetUsers")
    domain: str = Field(default="")
    tech_stack: ProjectTechStack = Field(alias="techStack")
    requirements: Requirements = Field(alias="requirements")
    features: Features
    ui_design: UIDesign = Field(alias="uiDesign")
    pages: Pages
    data_model: DataModel = Field(alias="dataModel")
    api: Api
    testing: Testing
    project_structure: ProjectStructure = Field(alias="projectStructure")
    deployment: Deployment
    documentation: Documentation

    class Config:
        populate_by_name = True


class ProjectTemplateResponse(BaseModel):
    """Response schema for a project template."""

    id: str
    template: ProjectTemplate


class ProjectTemplateList(BaseModel):
    """Response schema for a list of project templates."""

    templates: List[ProjectTemplateResponse] = Field(default_factory=list)

File: frontend/src/types/templates.ts
/**
 * Types for project templates.
 */

// Frontend Types
export interface FrontendTechStack {
  framework: string;
  language: string;
  stateManagement?: string;
  uiLibrary?: string;
  formHandling?: string;
  routing?: string;
  apiClient?: string;
  metaFramework?: string | null;
}

// Backend Types
export interface FrameworkBackend {
  type: "framework";
  framework: string; // Express.js, NestJS, Django, etc.
  language: string; // JavaScript, TypeScript, Python, etc.
  realtime?: string | null;
}

export interface BaaSBackend {
  type: "baas";
  service: string; // Supabase, Firebase, etc.
  functions?: string | null;
  realtime?: string | null;
}

export interface ServerlessBackend {
  type: "serverless";
  service: string; // AWS Lambda, Azure Functions, etc.
  language: string; // JavaScript, TypeScript, Python, etc.
}

// Combined Backend
export type BackendTechStack =
  | FrameworkBackend
  | BaaSBackend
  | ServerlessBackend;

// Database Types
export interface SQLDatabase {
  type: "sql";
  system: string; // PostgreSQL, MySQL, etc.
  hosting: string; // Supabase, AWS RDS, etc.
  orm?: string | null;
}

export interface NoSQLDatabase {
  type: "nosql";
  system: string; // MongoDB, Firestore, etc.
  hosting: string; // MongoDB Atlas, Firebase, etc.
  client?: string | null;
}

// Combined Database
export type DatabaseTechStack = SQLDatabase | NoSQLDatabase;

// Authentication
export interface AuthenticationTechStack {
  provider: string;
  methods: string[];
}

// Hosting
export interface HostingTechStack {
  frontend: string;
  backend: string;
  database?: string;
}

// Storage
export interface StorageTechStack {
  type: string; // objectStorage, fileSystem
  service: string;
}

// Deployment
export interface DeploymentTechStack {
  ci_cd?: string | null;
  containerization?: string | null;
}

// Complete Project Template
export interface ProjectTechStack {
  frontend: FrontendTechStack;
  backend: BackendTechStack;
  database: DatabaseTechStack;
  authentication: AuthenticationTechStack;
  hosting: HostingTechStack;
  storage?: StorageTechStack;
  deployment?: DeploymentTechStack;
}

export interface Requirements {
  functional: string[];
  non_functional: string[];
}

export interface FeatureModule {
  name: string;
  description: string;
  enabled: boolean;
  optional?: boolean;
  providers?: string[];
}

export interface Features {
  coreModules: FeatureModule[];
}

export interface PageComponent {
  name: string;
  path: string;
  components: string[];
  enabled: boolean;
}

export interface Pages {
  public: PageComponent[];
  authenticated: PageComponent[];
  admin: PageComponent[];
}

export interface EntityField {
  name: string;
  type: string;
  primaryKey?: boolean;
  generated?: boolean;
  unique?: boolean;
  required?: boolean;
  default?: string | number | boolean;
  enum?: string[];
  foreignKey?: Record<string, string | number | boolean>;
}

export interface Entity {
  name: string;
  description: string;
  fields: EntityField[];
}

export interface Relationship {
  type: string;
  from_entity: string;
  to_entity: string;
  field: string;
  throughTable?: string;
}

export interface DataModel {
  entities: Entity[];
  relationships: Relationship[];
}

export interface ApiParameter {
  name: string;
  type: string;
  required: boolean;
  description: string;
}

export interface ApiResponse {
  status: number;
  description: string;
  schema: string;
}

export interface ApiEndpoint {
  path: string;
  parameters?: ApiParameter[];
  description: string;
  methods: string[];
  auth: boolean;
  roles?: string[];
  responses?: ApiResponse[];
}

export interface Api {
  endpoints: ApiEndpoint[];
}

export interface TestingFramework {
  framework: string;
  coverage: number;
  directories?: string[];
  focus?: string[];
  scenarios?: string[];
}

export interface Testing {
  unit: TestingFramework;
  integration: TestingFramework;
  e2e: TestingFramework;
}

export interface ProjectStructure {
  frontend: {
    directories: string[];
    files: string[];
  };
}

export interface DeploymentVariable {
  name: string;
  value: string;
  secret: boolean;
}

export interface DeploymentEnvironment {
  name: string;
  url: string;
  variables: DeploymentVariable[];
}

export interface CICD {
  provider: string;
  steps: string[];
}

export interface Deployment {
  environments: DeploymentEnvironment[];
  cicd: CICD;
}

export interface Diagram {
  name: string;
  type: string;
  template: string;
}

export interface Documentation {
  sections: string[];
  diagrams: Diagram[];
}

export enum ImplementationPromptType {
  MAIN = "main",
  FOLLOWUP_1 = "followup_1",
  FOLLOWUP_2 = "followup_2",
}

export interface ImplementationPrompt {
  id: string;
  content: string;
  type: ImplementationPromptType;
  created_at?: string;
  updated_at?: string;
}

export interface ImplementationPrompts {
  data: Record<string, ImplementationPrompt[]>;
}

// UI Design Interfaces
export interface Colors {
  primary: string;
  secondary: string;
  accent: string;
  background: string;
  textPrimary: string;
  textSecondary: string;
  success: string;
  warning: string;
  error: string;
  info: string;
  surface: string;
  border: string;
}

export interface HeadingSizes {
  h1: string;
  h2: string;
  h3: string;
  h4: string;
  h5: string;
  h6: string;
}

export interface Typography {
  fontFamily: string;
  headingFont: string;
  fontSize: string;
  lineHeight: number;
  fontWeight: number;
  headingSizes: HeadingSizes;
}

export interface Spacing {
  unit: string;
  scale: number[];
}

export interface BorderRadius {
  small: string;
  medium: string;
  large: string;
  xl: string;
  pill: string;
}

export interface Shadows {
  small: string;
  medium: string;
  large: string;
  xl: string;
}

export interface Layout {
  containerWidth: string;
  responsive: boolean;
  sidebarWidth: string;
  topbarHeight: string;
  gridColumns: number;
  gutterWidth: string;
}

export interface Components {
  buttonStyle: string;
  inputStyle: string;
  cardStyle: string;
  tableStyle: string;
  navStyle: string;
}

export interface DarkModeColors {
  background: string;
  textPrimary: string;
  textSecondary: string;
  surface: string;
  border: string;
}

export interface DarkMode {
  enabled: boolean;
  colors: DarkModeColors;
}

export interface Animations {
  transitionDuration: string;
  transitionTiming: string;
  hoverScale: number;
  enableAnimations: boolean;
}

export interface UIDesign {
  colors: Colors;
  typography: Typography;
  spacing: Spacing;
  borderRadius: BorderRadius;
  shadows: Shadows;
  layout: Layout;
  components: Components;
  darkMode: DarkMode;
  animations: Animations;
}

export interface ProjectTemplate {
  id?: string;
  name: string;
  version: string;
  description: string;
  tags?: string[];
  businessGoals: string[];
  targetUsers: string;
  domain: string;
  techStack: ProjectTechStack;
  requirements: Requirements;
  features: Features;
  uiDesign: UIDesign;
  pages: Pages;
  dataModel: DataModel;
  api: Api;
  testing: Testing;
  projectStructure: ProjectStructure;
  deployment: Deployment;
  documentation: Documentation;
}

export interface ProjectTemplateResponse {
  id: string;
  template: ProjectTemplate;
}

export interface ProjectTemplateList {
  templates: ProjectTemplateResponse[];
}
